### PR TITLE
Feature/results system interface

### PR DESCRIPTION
**Pull request by collijk**
Ignore the wiring in the engine and plugins system. This is a documentation/API first approach to the results system.  Clearly the api documentation is going to be insufficient without usage examples.  In the short run, usage examples here: https://github.com/ihmeuw/vivarium_public_health/pull/224

### CODE REVIEW

**Review by KjellSwedin — APPROVED** [review]

File: src/vivarium/framework/results/manager.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    groupby-apply logic commonly done when manipulating :mod:`pandas`
+    DataFrames. The representation of state in the simulation is complex,
+    however, as it includes information both in the population state table
+    as well as dynamically generated information available from the
+    :class:`value pipelines <vivarium.framework.values.Pipeline>`.
+    Additionally, good encapsulation of simulation logic typically has
+    results production separated from the modelling code into specialized
+    `Observer` components. This often highlights the need for transformations
+    of the simulation state into representations that aren't needed for
+    modelling, but are required for the stratification of produced results.
```
Comment (mattkappel) [inline src/vivarium/framework/results/manager.py:36]: US English uses "modeling"

https://writingexplained.org/modelling-vs-modeling-difference

File: src/vivarium/framework/results/manager.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    of the expanded state table generated by the previously discussed mapping
+    strategies.
+
+    The suggested pattern for adding mapping strategies and default
+    grouping columns is to centralize them in the `setup` method of a general
+    `Metrics` or `Results` component. From this component you can expose
+    configuration for the global stratification of results. This component
+    should be fairly easy to reuse across entire categories of models.
+
+    Individual `Observer` components (or model components if modeling and
```
Comment (mattkappel) [inline src/vivarium/framework/results/manager.py:62]: Used "modeling" here.

**Review by mattkappel — APPROVED** [review]

File: src/vivarium/framework/engine.py [hunk 3]
```diff
@@ -184,13 +186,15 @@ def __init__(self, configuration, plugin_manager):
         self.value = plugin_manager.get_plugin_interface('value')                   # type: ValuesInterface
         self.event = plugin_manager.get_plugin_interface('event')                   # type: EventInterface
         self.population = plugin_manager.get_plugin_interface('population')         # type: PopulationInterface
-        self.resources = plugin_manager.get_plugin_interface('resource')             # type: ResourceInterface
+        self.resources = plugin_manager.get_plugin_interface('resource')            # type: ResourceInterface
+        self.results = plugin_manager.get_plugin_interface('results')               # type: ResultsInterface
```
Comment (rmudambi) [inline src/vivarium/framework/engine.py:190]: Need to add docs for this to match the api docs in my PR

File: src/vivarium/framework/results/__init__.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
+from vivarium.framework.results.manager import (
+    ResultsManager,
```
Comment (rmudambi) [inline src/vivarium/framework/results/__init__.py:2]: Why does the `ResultsManager` need to be exposed here?

File: src/vivarium/framework/results/manager.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    simulation state with transformations of the underlying state table and
+    pipeline values using :meth:`add_mapping_strategy`. For example, a public
+    health simulation may want to group its summary results by 5-year age
+    bins. A user could add a mapping strategy to transform the continuous
+    "age" property of each of the simulants into a discrete set of age
+    categories. This expanded version of the state can then be used to
+    stratify the population into groups of interest for producing aggregate
+    statistics.
+
+    In general, most of this grouping is for all results being produced.  E.g.
```
Comment (rmudambi) [inline src/vivarium/framework/results/manager.py:48]: "... for all produced results" reads more cleanly to me.

**Review by rmudambi — APPROVED** [review]

File: src/vivarium/framework/engine.py (hunk 3 above)
Comment (collijk) [reply src/vivarium/framework/engine.py:190]: I'll put it on my list for when the system is implemented.

File: src/vivarium/framework/results/__init__.py (hunk 4 above)
Comment (collijk) [reply src/vivarium/framework/results/__init__.py:2]: To match the interface of other system packages that are framework level modules.

File: src/vivarium/framework/results/manager.py (hunk 1 above)
Comment (collijk) [reply src/vivarium/framework/results/manager.py]: I've been reading too many lancet papers.

**Pull request merged**